File: library/src/main/java/it/sephiroth/android/library/tooltip/TooltipOverlayDrawable.java
package it.sephiroth.android.library.tooltip;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.view.animation.AccelerateDecelerateInterpolator;

/**
 * Created by alessandro on 12/12/15.
 */
public class TooltipOverlayDrawable extends Drawable {
    @SuppressWarnings ("unused")
    public static final String TAG = TooltipOverlay.class.getSimpleName();
    public static final float ALPHA_MAX = 255f;
    public static final double FADEOUT_START_DELAY = 0.55;
    public static final double FADEIN_DURATION = 0.3;
    public static final double SECOND_ANIM_START_DELAY = 0.25;
    private Paint mOuterPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private Paint mInnerPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private float mOuterRadius;
    private float mInnerRadius = 0;
    private AnimatorSet mFirstAnimatorSet;
    private AnimatorSet mSecondAnimatorSet;
    private ValueAnimator mFirstAnimator;
    private ValueAnimator mSecondAnimator;
    private int mRepeatIndex;
    private boolean mStarted;
    private int mOuterAlpha;
    private int mInnerAlpha;
    private int mRepeatCount = 1;
    private long mDuration = 400;

    public TooltipOverlayDrawable(Context context, int defStyleResId) {
        mOuterPaint.setStyle(Paint.Style.FILL);
        mInnerPaint.setStyle(Paint.Style.FILL);

        final TypedArray array =
            context.getTheme().obtainStyledAttributes(defStyleResId, R.styleable.TooltipOverlay);

        for (int i = 0; i < array.getIndexCount(); i++) {
            int index = array.getIndex(i);

            if (index == R.styleable.TooltipOverlay_android_color) {
                int color = array.getColor(index, 0);
                mOuterPaint.setColor(color);
                mInnerPaint.setColor(color);

            } else if (index == R.styleable.TooltipOverlay_ttlm_repeatCount) {
                mRepeatCount = array.getInt(index, 1);

            } else if (index == R.styleable.TooltipOverlay_android_alpha) {
                int alpha = (int) (array.getFloat(index, mInnerPaint.getAlpha() / ALPHA_MAX) * 255);
                mInnerPaint.setAlpha(alpha);
                mOuterPaint.setAlpha(alpha);

            } else if (index == R.styleable.TooltipOverlay_ttlm_duration) {
                mDuration = array.getInt(index, 400);
            }
        }

        array.recycle();

        mOuterAlpha = getOuterAlpha();
        mInnerAlpha = getInnerAlpha();

        // first
        Animator fadeIn = ObjectAnimator.ofInt(this, "outerAlpha", 0, mOuterAlpha);
        fadeIn.setDuration((long) (mDuration * FADEIN_DURATION));

        Animator fadeOut = ObjectAnimator.ofInt(this, "outerAlpha", mOuterAlpha, 0, 0);
        fadeOut.setStartDelay((long) (mDuration * FADEOUT_START_DELAY));
        fadeOut.setDuration((long) (mDuration * (1.0 - FADEOUT_START_DELAY)));

        mFirstAnimator = ObjectAnimator.ofFloat(this, "outerRadius", 0, 1);
        mFirstAnimator.setDuration(mDuration);

        mFirstAnimatorSet = new AnimatorSet();
        mFirstAnimatorSet.playTogether(fadeIn, mFirstAnimator, fadeOut);
        mFirstAnimatorSet.setInterpolator(new AccelerateDecelerateInterpolator());
        mFirstAnimatorSet.setDuration(mDuration);

        // second
        fadeIn = ObjectAnimator.ofInt(this, "innerAlpha", 0, mInnerAlpha);
        fadeIn.setDuration((long) (mDuration * FADEIN_DURATION));

        fadeOut = ObjectAnimator.ofInt(this, "innerAlpha", mInnerAlpha, 0, 0);
        fadeOut.setStartDelay((long) (mDuration * FADEOUT_START_DELAY));
        fadeOut.setDuration((long) (mDuration * (1.0 - FADEOUT_START_DELAY)));

        mSecondAnimator = ObjectAnimator.ofFloat(this, "innerRadius", 0, 1);
        mSecondAnimator.setDuration(mDuration);

        mSecondAnimatorSet = new AnimatorSet();
        mSecondAnimatorSet.playTogether(fadeIn, mSecondAnimator, fadeOut);
        mSecondAnimatorSet.setInterpolator(new AccelerateDecelerateInterpolator());
        mSecondAnimatorSet.setStartDelay((long) (mDuration * SECOND_ANIM_START_DELAY));
        mSecondAnimatorSet.setDuration(mDuration);

        mFirstAnimatorSet.addListener(new AnimatorListenerAdapter() {
            boolean cancelled;

            @Override
            public void onAnimationCancel(Animator animation) {
                super.onAnimationCancel(animation);
                cancelled = true;
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                if (!cancelled && isVisible() && ++mRepeatIndex < mRepeatCount) {
                    mFirstAnimatorSet.start();
                }
            }
        });

        mSecondAnimatorSet.addListener(new AnimatorListenerAdapter() {
            boolean cancelled;

            @Override
            public void onAnimationCancel(Animator animation) {
                super.onAnimationCancel(animation);
                cancelled = true;
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                if (!cancelled && isVisible() && mRepeatIndex < mRepeatCount) {
                    mSecondAnimatorSet.setStartDelay(0);
                    mSecondAnimatorSet.start();
                }
            }
        });

    }

    public int getOuterAlpha() {
        return mOuterPaint.getAlpha();
    }

    @SuppressWarnings ("unused")
    public void setOuterAlpha(final int value) {
        mOuterPaint.setAlpha(value);
        invalidateSelf();
    }

    public int getInnerAlpha() {
        return mInnerPaint.getAlpha();
    }

    @SuppressWarnings ("unused")
    public void setInnerAlpha(final int value) {
        mInnerPaint.setAlpha(value);
        invalidateSelf();
    }

    @Override
    public void draw(Canvas canvas) {
        Rect bounds = getBounds();
        int centerX = bounds.width() / 2;
        int centerY = bounds.height() / 2;
        canvas.drawCircle(centerX, centerY, mOuterRadius, mOuterPaint);
        canvas.drawCircle(centerX, centerY, mInnerRadius, mInnerPaint);

    }

    @Override
    public void setAlpha(int i) {

    }

    @Override
    public void setColorFilter(ColorFilter colorFilter) {

    }

    @Override
    public boolean setVisible(boolean visible, boolean restart) {
        boolean changed = isVisible() != visible;

        if (visible) {
            if (restart || !mStarted) {
                replay();
            }
        } else {
            stop();
        }

        return changed;
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSLUCENT;
    }

    @Override
    protected void onBoundsChange(Rect bounds) {
        super.onBoundsChange(bounds);
        mOuterRadius = Math.min(bounds.width(), bounds.height()) / 2;
        mFirstAnimator.setFloatValues(0, mOuterRadius);
        mSecondAnimator.setFloatValues(0, mOuterRadius);
    }

    @Override
    public int getIntrinsicWidth() {
        return 96;
    }

    @Override
    public int getIntrinsicHeight() {
        return 96;
    }

    public void play() {
        mRepeatIndex = 0;
        mStarted = true;
        mFirstAnimatorSet.start();
        mSecondAnimatorSet.setStartDelay((long) (mDuration * SECOND_ANIM_START_DELAY));
        mSecondAnimatorSet.start();
    }

    public void replay() {
        stop();
        play();
    }

    public void stop() {
        mFirstAnimatorSet.cancel();
        mSecondAnimatorSet.cancel();

        mRepeatIndex = 0;
        mStarted = false;

        setInnerRadius(0);
        setOuterRadius(0);
    }

    @SuppressWarnings ("unused")
    public float getInnerRadius() {
        return mInnerRadius;
    }

    @SuppressWarnings ("unused")
    public void setInnerRadius(final float rippleRadius) {
        mInnerRadius = rippleRadius;
        invalidateSelf();
    }

    @SuppressWarnings ("unused")
    public float getOuterRadius() {
        return mOuterRadius;
    }

    @SuppressWarnings ("unused")
    public void setOuterRadius(final float value) {
        mOuterRadius = value;
        invalidateSelf();
    }
}

File: xtooltip/src/main/java/it/sephiroth/android/library/xtooltip/TooltipOverlayDrawable.kt
package it.sephiroth.android.library.xtooltip


import android.animation.*
import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.view.animation.AccelerateDecelerateInterpolator
import timber.log.Timber

/**
 * Created by alessandro crugnola on 12/12/15.
 * alessandro.crugnola@gmail.com
 *
 *
 * LICENSE
 * Copyright 2015 Alessandro Crugnola
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy,
 * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software
 * is furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
 * OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

@Suppress("SpellCheckingInspection")
class TooltipOverlayDrawable(context: Context, defStyleResId: Int) : Drawable() {
    private val mOuterPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val mInnerPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var mOuterRadius: Float = 0.toFloat()
    private var innerRadius = 0f
        set(rippleRadius) {
            field = rippleRadius
            invalidateSelf()
        }
    private val mFirstAnimatorSet: AnimatorSet
    private val mSecondAnimatorSet: AnimatorSet
    private val mFirstAnimator: ValueAnimator
    private val mSecondAnimator: ValueAnimator
    private var mRepeatIndex: Int = 0
    private var mStarted: Boolean = false
    private val mOuterAlpha: Int
    private val mInnerAlpha: Int
    private var mRepeatCount = 1
    private var mDuration: Long = 400

    private var outerAlpha: Int
        get() = mOuterPaint.alpha
        set(value) {
            mOuterPaint.alpha = value
            invalidateSelf()
        }

    private var innerAlpha: Int
        get() = mInnerPaint.alpha
        set(value) {
            mInnerPaint.alpha = value
            invalidateSelf()
        }

    private var outerRadius: Float
        get() = mOuterRadius
        set(value) {
            mOuterRadius = value
            invalidateSelf()
        }

    init {
        mOuterPaint.style = Paint.Style.FILL
        mInnerPaint.style = Paint.Style.FILL

        val array = context.theme.obtainStyledAttributes(defStyleResId, R.styleable.TooltipOverlay)

        for (i in 0 until array.indexCount) {
            val index = array.getIndex(i)

            when (index) {
                R.styleable.TooltipOverlay_android_color -> {
                    val color = array.getColor(index, 0)
                    mOuterPaint.color = color
                    mInnerPaint.color = color

                }
                R.styleable.TooltipOverlay_ttlm_repeatCount -> mRepeatCount = array.getInt(index, 1)
                R.styleable.TooltipOverlay_android_alpha -> {
                    val alpha = (array.getFloat(index, mInnerPaint.alpha / ALPHA_MAX) * 255).toInt()
                    mInnerPaint.alpha = alpha
                    mOuterPaint.alpha = alpha

                }
                R.styleable.TooltipOverlay_ttlm_duration -> mDuration = array.getInt(index, 400).toLong()
            }
        }

        array.recycle()

        mOuterAlpha = outerAlpha
        mInnerAlpha = innerAlpha

        // first
        var fadeIn: Animator = ObjectAnimator.ofInt(this, "outerAlpha", 0, mOuterAlpha)
        fadeIn.duration = (mDuration * FADEIN_DURATION).toLong()

        var fadeOut: Animator = ObjectAnimator.ofInt(this, "outerAlpha", mOuterAlpha, 0, 0)
        fadeOut.startDelay = (mDuration * FADEOUT_START_DELAY).toLong()
        fadeOut.duration = (mDuration * (1.0 - FADEOUT_START_DELAY)).toLong()

        mFirstAnimator = ObjectAnimator.ofFloat(this, "outerRadius", 0f, 1f)
        mFirstAnimator.duration = mDuration

        mFirstAnimatorSet = AnimatorSet()
        mFirstAnimatorSet.playTogether(fadeIn, mFirstAnimator, fadeOut)
        mFirstAnimatorSet.interpolator = AccelerateDecelerateInterpolator()
        mFirstAnimatorSet.duration = mDuration

        // second
        fadeIn = ObjectAnimator.ofInt(this, "innerAlpha", 0, mInnerAlpha)
        fadeIn.duration = (mDuration * FADEIN_DURATION).toLong()

        fadeOut = ObjectAnimator.ofInt(this, "innerAlpha", mInnerAlpha, 0, 0)
        fadeOut.setStartDelay((mDuration * FADEOUT_START_DELAY).toLong())
        fadeOut.duration = (mDuration * (1.0 - FADEOUT_START_DELAY)).toLong()

        mSecondAnimator = ObjectAnimator.ofFloat(this, "innerRadius", 0f, 1f)
        mSecondAnimator.duration = mDuration

        mSecondAnimatorSet = AnimatorSet()
        mSecondAnimatorSet.playTogether(fadeIn, mSecondAnimator, fadeOut)
        mSecondAnimatorSet.interpolator = AccelerateDecelerateInterpolator()
        mSecondAnimatorSet.startDelay = (mDuration * SECOND_ANIM_START_DELAY).toLong()
        mSecondAnimatorSet.duration = mDuration

        mFirstAnimatorSet.addListener(object : AnimatorListenerAdapter() {
            var cancelled: Boolean = false

            override fun onAnimationCancel(animation: Animator) {
                super.onAnimationCancel(animation)
                cancelled = true
            }

            override fun onAnimationEnd(animation: Animator) {
                if (!cancelled && isVisible && ++mRepeatIndex < mRepeatCount) {
                    mFirstAnimatorSet.start()
                }
            }
        })

        mSecondAnimatorSet.addListener(object : AnimatorListenerAdapter() {
            var cancelled: Boolean = false

            override fun onAnimationCancel(animation: Animator) {
                super.onAnimationCancel(animation)
                cancelled = true
            }

            override fun onAnimationEnd(animation: Animator) {
                if (!cancelled && isVisible && mRepeatIndex < mRepeatCount) {
                    mSecondAnimatorSet.startDelay = 0
                    mSecondAnimatorSet.start()
                }
            }
        })

    }

    override fun draw(canvas: Canvas) {
        val bounds = bounds
        val centerX = bounds.width() / 2
        val centerY = bounds.height() / 2
        canvas.drawCircle(centerX.toFloat(), centerY.toFloat(), mOuterRadius, mOuterPaint)
        canvas.drawCircle(centerX.toFloat(), centerY.toFloat(), innerRadius, mInnerPaint)

    }

    override fun setAlpha(i: Int) {}

    override fun setColorFilter(colorFilter: ColorFilter?) {}

    override fun setVisible(visible: Boolean, restart: Boolean): Boolean {
        val changed = isVisible != visible

        if (visible) {
            if (restart || !mStarted) {
                replay()
            }
        } else {
            stop()
        }

        return changed
    }

    override fun getOpacity(): Int {
        return PixelFormat.TRANSLUCENT
    }

    override fun onBoundsChange(bounds: Rect) {
        Timber.i("onBoundsChange: $bounds")
        super.onBoundsChange(bounds)
        mOuterRadius = (Math.min(bounds.width(), bounds.height()) / 2).toFloat()
        mFirstAnimator.setFloatValues(0f, mOuterRadius)
        mSecondAnimator.setFloatValues(0f, mOuterRadius)
    }

    override fun getIntrinsicWidth(): Int {
        return 96
    }

    override fun getIntrinsicHeight(): Int {
        return 96
    }

    private fun play() {
        mRepeatIndex = 0
        mStarted = true
        mFirstAnimatorSet.start()
        mSecondAnimatorSet.startDelay = (mDuration * SECOND_ANIM_START_DELAY).toLong()
        mSecondAnimatorSet.start()
    }

    private fun replay() {
        stop()
        play()
    }

    private fun stop() {
        mFirstAnimatorSet.cancel()
        mSecondAnimatorSet.cancel()

        mRepeatIndex = 0
        mStarted = false

        innerRadius = 0f
        outerRadius = 0f
    }

    @Suppress("SpellCheckingInspection")
    companion object {
        const val ALPHA_MAX = 255f
        const val FADEOUT_START_DELAY = 0.55
        const val FADEIN_DURATION = 0.3
        const val SECOND_ANIM_START_DELAY = 0.25
    }
}
